### PR TITLE
WT-12331 API review

### PR DESCRIPTION
**Pull request by marcbutler**
Add the header only for API review.

 - For API design review ONLY.
 - Not in WT normal style: will be changed once API design is agreed.

### CODE REVIEW

File: src/include/wtfutex.h [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  WT_WAKE_ALL			/* Wake all waiting threads. */
+};
+
+
+/*
+ * Get the value associated with the futex.
+ *
+ * \retval -EINVAL Not a futex, or pointer to [val] is null.
+ */
+int __wt_fetch(struct __wt_futex *ftx, uint32_t *val);
```
Comment (agorrod) [inline src/include/wtfutex.h:27]: Is there a reason not to call this __wt_futex_fetch? I'm trying to figure out if fetch/store consciously skip the `futex` prefix.

I'm not that familiar with futexes, but the interface seems sensible to me as is.

Comment (marcbutler) [reply src/include/wtfutex.h]: No: entirely inadvertent. Fixed -- thanks.

**Review by keitharnoldsmith — APPROVED** [review]
I started a review, got interrupted, ate lunch, came back... and found that Peter had stolen my comment about negative error returns! :-)

I'm not sure whether the header name is intended to be part of the API up for review here, but I'd call this `os_futex.h` since it's abstracting functionality across multiple operating systems, or possibly just place this content in `os.h`.

I'll LGTM and follow along on Peter's comments.

File: src/include/wtfutex.h [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * \retval -EINVAL Not a futex.
+ */
+int __wt_futex_store(struct __wt_futex *ftx, uint32_t val);
+
+
+/*
+ * Wait on a futex.
+ *
+ * \param expected Expected current value of futex.
```
[the remark this answers is not in the document]
Comment (marcbutler) [reply src/include/wtfutex.h]: It is treated as a wakeup.

File: src/include/wtfutex.h [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * \retval -EINVAL Not a futex, or [timeout_us] <= 0.
+ * \retval -ETIMEDOUT Timeout expired before being awoken.
```
[the remark this answers is not in the document]
Comment (marcbutler) [reply src/include/wtfutex.h]: If the futex call is interrupted, a new timeout is calculated and it goes back to waiting.

File: src/include/wtfutex.h [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+ * Thread
+ *  T1 : __wt_wait(F, X, ALONGTIME);
+ *  T2 : __wt_store(F, Y);
+ *  T1 : __wt_wait(F, X, ALONGTIME) == 0 "spurious wakeup"
+ *
+ * Thread T1 will see this as an intentional wakeup.
+ *
+ * \retval -EINVAL Not a futex.
+ */
+int __wt_futex_store(struct __wt_futex *ftx, uint32_t val);
```
[the remark this answers is not in the document]
Comment (marcbutler) [reply src/include/wtfutex.h]: The idea here originally was to provide some kind of wrapper for a basic use: with the idea of trying to create a usable abstract mechanism. However you raise a good point: this is a low-level mechanism and we will likely want to vary how we modify the associated 32-bit value. So I have pared the API to a portable "futex ops" API.

**Pull request closed without merging**